ccl/sqlproxyccl: developer provided proxy handler

darinpp opened the pull request:
This PR changes the sqlproxyccl's Proxy implementation. Previously, it
was expected that the developer will populate an options struct that
will customize how the provided proxy will operate. With this PR, it is
expected that the developer will provide a proxy handler instead that
will do everything needed to proxy an SQL connection from a frontend to
a backend. To help with the implementation, the code has been split into
number of helper methods that can be combined and that make easy bulding
a custom proxy. The existing implementation is now moved into the test
file and is just one example of how a complicated proxy can be
constructed.

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/58203)
<!-- Reviewable:end -->

andy-kimball reviewed (commented):
Should we call this `Handler` rather than `Interceptor`? I think `Handler` is the usual Go term for what this is doing.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58203#-:-MPBSjXkDkhHgSK24eSn:b-1ejw1z)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball and @darinpp)

___
*[pkg/ccl/sqlproxyccl/proxy.go, line 223 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58203#-MPBT9nDBpgMTNibnmrT:-MPBT9nDBpgMTNibnmrU:b-jny6uq) ([raw file](https://github.com/cockroachdb/cockroach/blob/122585321d3fb0b1ee21bf4613870f5392758136/pkg/ccl/sqlproxyccl/proxy.go#L223)):*
> ```Go
> 	}
> 
> 	go func() {
> ```

Why are we introducing another goroutine here? I'd like to find a design that doesn't require an extra goroutine for every connection.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58203)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @darinpp)

___
*[pkg/ccl/sqlproxyccl/proxy.go, line 83 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58203#-MPBgpah7DurPww0whPB:-MPBgpah7DurPww0whPC:b-fw251s) ([raw file](https://github.com/cockroachdb/cockroach/blob/f38d532630ba36085356ba7bc0d83143046ca3c3/pkg/ccl/sqlproxyccl/proxy.go#L83)):*
> ```Go
> 	// is accepted. It can optionally negotiate SSL, provide admittance control or
> 	// other types of frontend connection filtering.
> 	FrontendAdmitter func(incoming net.Conn) (net.Conn, *pgproto3.StartupMessage, error)
> ```

Let's get rid of `FrontendAdmitter` and `BackendDialer`, since they're obsolete with the new proxy pattern.
___
*[pkg/ccl/sqlproxyccl/proxy.go, line 96 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58203#-MPBgbHS4fbSjTGsFfzE:-MPBgbHS4fbSjTGsFfzF:b-oabw61) ([raw file](https://github.com/cockroachdb/cockroach/blob/f38d532630ba36085356ba7bc0d83143046ca3c3/pkg/ccl/sqlproxyccl/proxy.go#L96)):*
> ```Go
> // information to the end user in case of a problem.
> func (s *Server) SendErrToClient(conn net.Conn, code ErrorCode, msg string) {
> 	if s.opts.OnSendErrToClient != nil {
> ```

Let's get rid of `OnSendErrToClient`, since now the caller can do that on their own if they need it.
___
*[pkg/ccl/sqlproxyccl/proxy.go, line 233 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58203#-MPBiaRGErALBlhF_VlY:-MPBiaRGErALBlhF_VlZ:bnjiuz1) ([raw file](https://github.com/cockroachdb/cockroach/blob/f38d532630ba36085356ba7bc0d83143046ca3c3/pkg/ccl/sqlproxyccl/proxy.go#L233)):*
> ```Go
> }
> 
> func (s *Server) updateMetricsForError(err error) {
> ```

How come this is not exported? Wouldn't hosts need to use it?


<!-- Sent from Reviewable.io -->

darinpp committed:
ccl/sqlproxyccl: developer provided proxy handler

This PR changes the sqlproxyccl's Proxy implementation. Previously, it
was expected that the developer will populate an options struct that
will customize how the provided proxy will operate. With this PR, it is
expected that the developer will provide a proxy handler instead that
will do everything needed to proxy an SQL connection from a frontend to
a backend. To help with the implementation, the code has been split into
number of helper methods that can be combined and that make easy bulding
a custom proxy. The existing implementation is now moved into the test
file and is just one example of how a complicated proxy can be
constructed.

Release note: none

Release justification:
